src/main/java/bio/terra/service/dataset/DatasetDao.java

```diff
@@ -211,15 +177,116 @@ private DataAccessException getFaultToInsert() {
+        logger.debug("Unlocking shared lock for datasetId: {}, flightId: {}", datasetId, flightId);
+
+        int numRowsUpdated = performLockQuery(sql, params, LockType.UnlockShared, null);
+
+        boolean unlockSucceeded = (numRowsUpdated == 1);
+        logger.info("Lock Operation: Unlock shared successful? {}; for datasetId: {}, flightId: {}",
+            unlockSucceeded, datasetId, flightId);
+        return unlockSucceeded;
+    }
+
+    private enum LockType {
+        LockExclusive,
+        LockShared,
+        UnlockExclusive,
+        UnlockShared
+    }
+    private int performLockQuery(String sql, MapSqlParameterSource params,
+                            LockType lockType, UUID datasetId) {
+        int numRowsUpdated = 0;
+        DataAccessException faultToInsert = getFaultToInsert(lockType);
+        try {
+            if (faultToInsert != null) {
+                logger.info("Test: Inserting fault to lock/unlock operation.");
+                throw faultToInsert;
+            }
+
+            numRowsUpdated = jdbcTemplate.update(sql, params);
+
+            if (numRowsUpdated == 0 &&
+                (lockType.equals(LockType.LockExclusive) || lockType.equals(LockType.LockShared))) {
+                // this method checks if the dataset exists
+                // if it does not exist, then the method throws a DatasetNotFoundException
+                // we don't need the result (dataset summary) here, just the existence check,
+                // so ignore the return value.
+                retrieveSummaryById(datasetId);
+
+                throw new DatasetLockException("Failed to lock the dataset");
+            }
+        } catch (DatasetNotFoundException notFound) {
+            logger.error("Dataset lock failed: Dataset not found. Lock Type: {}, DatasetId: {}", lockType, datasetId);
+            throw notFound;
+        } catch (DatasetLockException lockException) {
+            logger.error("Dataset lock failed: Failed to lock dataset. Lock Type: {}, DatasetId: {}",
+                lockType, datasetId);
+            throw lockException;
+        } catch (DataAccessException dataAccessException) {
+            if (retryQuery(dataAccessException)) {
+                logger.error("Dataset lock failed with retryable exception. Lock Type: {}, DatasetId: {}",
+                    lockType, datasetId);
+                throw new RetryQueryException("Retry", dataAccessException);
+            }
+            logger.error("Dataset lock failed with fatal exception. Lock Type: {}, DatasetId: {}",
+                lockType, datasetId);
+            throw dataAccessException;
+        }
+
         logger.debug("numRowsUpdated=" + numRowsUpdated);
-        return (numRowsUpdated == 1);
+        return numRowsUpdated;
+    }
+
+    private DataAccessException getFaultToInsert(LockType lockType) {
+        ConfigEnum RetryableFault = null;
+        ConfigEnum FatalFault = null;
+
+        switch (lockType) {
```

Comment:
I wonder if we need so many different   faults. The reason to have them is if we make tests that set more than one.
When do you think?

Reply:
This is a really good point, thank you. 
 I originally thought they needed to be separate, but I think I can combine the exclusive/shared lock faults.

The way I have my tests set up, I can't really distinguish between the period of time that the flight is attempting a lock vs. an unlock.  I just insert the fault at the beginning and remove it part of the way through the run.  I also still need to distinguish between the retry and fatal faults. However, I can specifically insert the faults during a shared or an exclusive locking flight, so I don't need distinguishing faults for those.